app/services/platform-apps/container-manager.ts

```diff
@@ -0,0 +1,360 @@
+    // teardown.
+    if (transform.electronWindowId === electronWindowId) {
+      cont.transform.next({
+        ...transform,
+        mounted: false,
+        electronWindowId: null,
+        slobsWindowId: null,
+      });
+
+      if (!cont.persistent) {
+        this.destroyContainer(containerId);
+      }
+    }
+  }
+
+  /**
+   * Refreshes all currently running containers for the provided app.
+   * @param app The app to refresh
+   */
+  refreshContainers(app: ILoadedApp) {
+    this.containers.forEach(cont => {
+      if (cont.appId === app.id) cont.container.webContents.reload();
+    });
+  }
+
+  private getContainerInfoForSlot(app: ILoadedApp, slot: EAppPageSlot): IContainerInfo {
+    const existingContainer = this.containers.find(
+      cont => cont.appId === app.id && cont.slot === slot,
+    );
+
+    if (existingContainer) {
+      return existingContainer;
+    }
+
+    return this.createContainer(app, slot);
+  }
+
+  private createContainer(app: ILoadedApp, slot: EAppPageSlot, persistent = false): IContainerInfo {
+    const view = new electron.remote.BrowserView({
+      webPreferences: {
+        nodeIntegration: false,
+        partition: this.getAppPartition(app),
+        preload: path.resolve(electron.remote.app.getAppPath(), 'bundles', 'guest-api'),
+      },
+    });
+
+    const info: IContainerInfo = {
+      slot,
+      persistent,
+      container: view,
+      appId: app.id,
+      transform: new BehaviorSubject<IBrowserViewTransform>({
+        pos: { x: 0, y: 0 },
+        size: { x: 0, y: 0 },
+        mounted: false,
+        electronWindowId: null,
+        slobsWindowId: null,
+      }),
+    };
+
+    if (app.unpacked) view.webContents.openDevTools();
+
+    view.webContents.on('did-finish-load', () => {
+      this.exposeApi(app, view.webContents.id, info.transform);
+    });
+
+    /**
+     * This has to be done from the main process to work properly
+     * @see https://github.com/electron/electron/issues/1378
+     */
+    electron.ipcRenderer.send('webContents-preventNavigation', view.webContents.id);
+
+    // We allow opening dev tools for beta apps only
+    if (app.beta) {
+      view.webContents.on('before-input-event', (e, input) => {
+        if (input.type === 'keyDown' && input.code === 'KeyI' && input.control && input.shift) {
+          view.webContents.openDevTools();
+        }
+      });
+    }
+
+    view.webContents.loadURL(this.getPageUrlForSlot(app, slot));
+
+    this.containers.push(info);
+
+    return info;
+  }
+
+  private destroyContainer(containerId: number) {
+    const cont = this.containers.find(cont => cont.container.id === containerId);
+    this.containers = this.containers.filter(c => c.container.id !== containerId);
+
+    // Electron types are incorrect here.  This method exists and is documented, but
+    // does not appear in the type definitions.
+    (cont.container as any).destroy();
```

Comment:
Can we ensure we'll always get a result from the `find` call above? Otherwise, that'll blow on `undefined`.